### PR TITLE
[TTAHUB-2439] Grant replacement: models and migrations

### PR DESCRIPTION
**Pull request by nvms**
## Description of change

Support for 1->n and n->1 grant replacements by introducing three new tables:

- `GrantReplacements`
- `GrantReplacementTypes`
- `GrantRelationshipToActive`

These are fairly straight-forward constructs, so looking at the models should tell you everything you need to know about them and how they are intended to be used.

The bulk of the changes are in

- `processFiles` in `src/lib/updateGrantsRecipients.js`
- `mergeGoals` in `src/goalServices/goals.js`
- various handlers/services that query the `Grants` table or referenced `oldGrantId`

The changes to the `Grant` `filtersToScopes` are not strictly necessary any longer since we decided to keep the `inactivationDate` and `inactivationReason` (this meant no join was necessary), but I left it there because I think it could be useful in the future. For example, if we wanted to create a new filter that returns all grants that a particular grant has replaced, we could do so without writing any raw SQL in the filter by leveraging the `return { includes: [] }` pattern.

### inactivationDate and inactivationReason

It seems we need to keep these because they *are not the same* as `replacementDate` and `replacementType`. For example:

This is an excerpt from `grant_award.xml`:

```xml
  <grant_award_id>7842</grant_award_id>
  ...
  <grant_status>Inactive</grant_status>
  <inactivation_date>2022-07-31</inactivation_date>
  <inactivation_reason>Replaced</inactivation_reason>
```

And from `grant_award_replacement.xml` for the same grant (7842):

```xml
<grant_award_replacement>
    <replaced_grant_award_id>7842</replaced_grant_award_id>
    <replacement_grant_award_id>8110</replacement_grant_award_id>
    <replacement_date>10-01-2021</replacement_date>
    <grant_replacement_type>DRS Non-Competitive Continuation</grant_replacement_type>
</grant_award_replacement>
```

Note the inactivation date is not the same as the replacement date, and the inactivation reason is not the same as the replacement type.

`inactivation_reason` -> `Grant.inactivationReason`
`grant_replacement_type` -> unique `GrantReplacementTypes.name`

`inactivation_date` -> `Grant.inactivationDate`
`replacement_date` -> `GrantReplacement.replacementDate`

New tests were added for the `processFiles` method to validate this.

## How to test

- Walk through the logic in `src/lib/updateGrantsRecipients.js` for the `processFiles` function. This still populates columns on `Grant` like `inactivationDate` and `inactivationReason`, but it now also creates `GrantReplacement` and `GrantReplacementType` entities. After an import, check your database to make sure these rows are created.
- There's no changes to the UI, so the only tests we can do are make sure everything still works, which I think the e2e tests do a pretty good job of.
- Validate the changes in `src/goalServices/goals.js` for the `mergeGoals` function. It should grab all old goals from all grants being replaced, not just the one previously referenced by the now-removed `oldGrantId` column.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-2439
* https://jira.acf.gov/browse/TTAHUB-2440
* https://jira.acf.gov/browse/TTAHUB-2441
* https://jira.acf.gov/browse/TTAHUB-2442
* https://jira.acf.gov/browse/TTAHUB-2443
* https://jira.acf.gov/browse/TTAHUB-2444
* https://jira.acf.gov/browse/TTAHUB-2445

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: README.md
Comment (thewatermethod) [inline README.md:93]: 🔥

File: src/lib/updateGrantsRecipients.js
Comment (thewatermethod) [inline src/lib/updateGrantsRecipients.js:412]: we've been removing instances of findOrCreate from our codebase since (I believe, @GarrettEHill might know better) the creates run outside of transactions.

**Review by thewatermethod — APPROVED** [review]
It all looks really good to me, ran through the suggested code spots and checked from the UI and can't find any issues. Can we wait until another review gets through before merging?

File: src/services/activityReports.js
Comment (kryswisnaskas) [inline src/services/activityReports.js:654]: Should we specify the attributes? (same comment for line 649 and 644)

File: src/goalServices/getGoalIdsBySimiliarity.test.js
Comment (AdamAdHocTeam) [inline src/goalServices/getGoalIdsBySimiliarity.test.js:5]: Should this test have something were we expect goals linked to older replaced grants to also come back? Or maybe that's in another test I haven't gotten to yet.

File: src/lib/updateGrantsRecipients.js
Comment (GarrettEHill) [reply src/lib/updateGrantsRecipients.js:412]: yes please do not use findOrCreate its a mess

Comment (AdamAdHocTeam) [reply src/lib/updateGrantsRecipients.js:412]: Yea I would double check this with Garrett.

File: src/scopes/grants/activeWithin.js
Comment (AdamAdHocTeam) [inline src/scopes/grants/activeWithin.js:67]: Are these duplicated?

File: src/services/recipient.js
Comment (AdamAdHocTeam) [inline src/services/recipient.js:129]: is this also duplicated?

File: src/widgets/helpers.js
Comment (AdamAdHocTeam) [inline src/widgets/helpers.js:71]: should we put this date in a const.

**Review by AdamAdHocTeam — APPROVED** [review]
Overall LGTM added some minor comments about duplicate code.

File: src/migrations/20241003202817-populate-grant-replacements.js
Comment (kryswisnaskas) [reply src/migrations/20241003202817-populate-grant-replacements.js:112]: Got it, this makes sense for backfilling historical data in this migration. Just to note, for future cases, we should avoid using `createdAt `and `updatedAt `for arbitrary values and stick to actual create/update operations.

File: src/scopes/grants/activeWithin.js
Comment (nvms) [reply src/scopes/grants/activeWithin.js:67]: No, because we  want either null or gte the provided date:

```ts
  [Op.or]: [
    { inactivationDate: { [Op.gte]: new Date(sd) } },
    { inactivationDate: null },
  ],
```

File: src/goalServices/goals.js
Comment (nvms) [reply src/goalServices/goals.js:2006]: Do you mean these two areas?

```js
// line 1988
  const grantsWithReplacementsDictionary = {};

  grantsWithReplacements.forEach((grant) => {
    if (grant.status === 'Active') {
      if (Array.isArray(grantsWithReplacementsDictionary[grant.id])) {
        grantsWithReplacementsDictionary[grant.id].push(grant.id);
      } else {
        grantsWithReplacementsDictionary[grant.id] = [grant.id];
      }
    } else {
      grant.grantRelationships.forEach((relationship) => {
        if (relationship.activeGrant && relationship.activeGrant.status === 'Active') {
          if (Array.isArray(grantsWithReplacementsDictionary[grant.id])) {
            grantsWithReplacementsDictionary[grant.id].push(relationship.activeGrantId);
          } else {
            grantsWithReplacementsDictionary[grant.id] = [relationship.activeGrantId];
          }
        }
      });
    }
  });
```

```js
// line 1694
  const grantLookup = {};
  grants.forEach((grant) => {
    if (grant.status === 'Active') {
      grantLookup[grant.id] = grant.id;
    } else {
      grant.grantRelationships.forEach((relationship) => {
        if (relationship.activeGrant && relationship.activeGrant.status === 'Active') {
          grantLookup[grant.id] = relationship.activeGrantId;
        }
      });
    }
  });
```

They look similar but do different things. I don't think a helper makes a lot of sense here

Comment (nvms) [reply src/goalServices/goals.js]: Good idea! Done here: 08046ccf4e878c6a3b4cf75a9347bcaf3e370fcc

**Pull request merged**
